Add JTDX Support

SydneyOwl opened the pull request:
Hello @xylo04,

Thank you for sharing this awesome project with us!

The current code seems to have decoding errors for JTDX messages:
<img width="500"  alt="error" src="https://github.com/user-attachments/assets/35e6afde-e0ac-4f24-8d15-44275ac268e7" />

After reviewing the [JTDX source code](https://github.com/jtdx-project/jtdx/blob/master/NetworkMessage.hpp), I noticed that its message protocol is very similar to WSJT-X's. It seems that with a few adjustments to the existing code, we could add support for JTDX!

Here is a demonstration of the decoding working correctly after the fix:
https://github.com/user-attachments/assets/f76462b5-2713-4f29-b35b-b1e0727da80a

I've submitted the PR with proposed fix - would you mind taking a look when you have a moment?

Wishing you a Happy New Year!
SydneyOwl

SydneyOwl committed:
JTDX Support

SydneyOwl committed:
Add comment

xylo04 commented:
Hi @SydneyOwl! This looks reasonably unobtrusive and in line with the spirit of the current implementation. I worry what would happen if WSJTX and JTDX evolve the wire protocol in incompatible ways, but I guess that's a bridge we can cross when we get to it.

Do you mind adding one or two JTDX test cases in `integration/receive_test.go` to make sure we don't break it in the future?

SydneyOwl committed:
Added JTDX test cases

SydneyOwl commented:
Hello @xylo04,

I have added some test cases to cover JTDX messages. Based on the source code, JTDX’s message format differs from WSJT-X only in four types: Heartbeat, Status, QSOLogged, and WSPRDecode. Accordingly, I have included test cases specifically for these four message types and all new tests are passing locally.

 Let me know if any adjustments are needed!

SydneyOwl committed:
Removed redundant fields from JTDX tests

SydneyOwl committed:
Merge branch 'v4' into v4

SydneyOwl commented:
Hi @xylo04,

Would you mind checking out the tests when you got any time? Thanks!

xylo04 commented on integration/receive_test.go:
```diff
@@ -133,6 +169,30 @@ func (s *integrationTestSuite) TestReceiveCases() {
+			want: receiveResult{
+				msg: wsjtx.QsoLoggedMessage{
+					Id:             "JTDX",
+					DateTimeOff:    parseTime("2026-01-05 12:31:00 +0800 CST"),
```
This is failing in CI but only because of the time zone. Convert to UTC to fit the test expectations
```suggestion
					DateTimeOff:    parseTime("2026-01-05 04:31:00 +0000 UTC"),
```

xylo04 commented on integration/receive_test.go:
```diff
@@ -133,6 +169,30 @@ func (s *integrationTestSuite) TestReceiveCases() {
+					TxPower:        "20",
+					Comments:       "Distance: 10773 km",
+					Name:           "Smith",
+					DateTimeOn:     parseTime("2026-01-05 12:29:47 +0800 CST"),
```
```suggestion
					DateTimeOn:     parseTime("2026-01-05 04:29:47 +0000 UTC"),
```

xylo04 reviewed (commented):
My apologies for the delayed response. Thanks for adding tests! They look good but failed in CI.

xylo04 committed:
Change tests from UTC+8 to UTC

Should pass in CI after that.

xylo04 commented:
Wait, it just occurred to me: *I* was able to commit directly to *your* fork's branch? Weird... Anyway, that still failed again in CI because it's expecting the timezone "UTC" and instead getting "Local", so I'm cloning it locally for a closer look.

xylo04 commented on integration/receive_test.go:
```diff
@@ -133,6 +169,30 @@ func (s *integrationTestSuite) TestReceiveCases() {
+			want: receiveResult{
+				msg: wsjtx.QsoLoggedMessage{
+					Id:             "JTDX",
+					DateTimeOff:    parseTime("2026-01-05 04:31:00 +0000 UTC"),
```
```suggestion
					DateTimeOff:    time.Date(2026, time.January, 5, 12, 31, 0, 0, time.Local),
```

xylo04 commented on integration/receive_test.go:
```diff
@@ -133,6 +169,30 @@ func (s *integrationTestSuite) TestReceiveCases() {
+					TxPower:        "20",
+					Comments:       "Distance: 10773 km",
+					Name:           "Smith",
+					DateTimeOn:     parseTime("2026-01-05 04:29:47 +0000 UTC"),
```
```suggestion
					DateTimeOn:     time.Date(2026, time.January, 5, 12, 29, 47, 0, time.Local),
```

xylo04 committed:
Use Local timezone

JTDX sends its date times over the wire with [`QDateTime::timeSpec`](https://doc.qt.io/qt-6/qdatetime.html#timeSpec) of 0 for local, unzoned date-times. Rather than trying to make any assumptions, this binding will pass that through.

xylo04 commented:
Leaving aside the fact that I'm able to commit code review suggestions to your fork... I've fixed this by building the expected dates with the `time.Date()` function and `time.Local` zone. This highlights a potential pitfall that JTDX is apparently sending zoneless times where WSJTX was sending zoned times. Keep that in mind while using the library.

Thanks for your contribution!
